[PCompiler] Correct foreach -> while transformation

aman-goel opened the pull request:
[PCompiler]
Corrects foreach to while transformation to account for possible continue statement in loop body

[Tst]
Adds regression test for foreach with continue

ankushdesai reviewed (changes requested):
At some point the code generation in the C# was changed from `foreach` statement to this `while` loop logic. I would like to know why?

P foreach can be simply converted to C# `foreach` can we roll back to that change?

Also, can you add more regressions both positive and negative.
For break, continue and combination of break and continue used at the same time.

aman-goel commented:
> At some point the code generation in the C# was changed from `foreach` statement to this `while` loop logic. I would like to know why?
> 
> P foreach can be simply converted to C# `foreach` can we roll back to that change?
> 
> Also, can you add more regressions both positive and negative. For break, continue and combination of break and continue used at the same time.

We added foreach to while IR transformation in #500 to support foreach in PSym/PCover. Originally, it was done only for PSym backend as a language/backend specific transformation. In the PR #500, based on the suggested comments, we enabled this transformation to all backends (including for PObserve) via 1b33a4628d38b61cc2deb68ec6314925fa72855b.

There was a minor bug in this transformation when the loop body has a continue statement (which this PR corrects).

Also added more +/- regression tests for foreach with break/continue via 82d768e681c27b6bde484d9cc2c78e05f881501f.
